text/0001-script-based-tokenizer.md

```diff
@@ -0,0 +1,318 @@
+
+```rust
+pub struct AnalyzerConfig<'a, A> {
+    /// language specialized pipeline, this can be switched during
+    /// document tokenization if the document contains several languages
+    pub pipeline_map: HashMap<(Script, Language), Pipeline>,
+    pub stop_words: &'a Set<A>,
+}
+```
+
+#### Analyzer
+
+The analyzer exposes an abstracted interface to the tokenization process. Its API is standard:
+
+```rust
+use crate::token::Token;
+use crate::tokenizer::Tokenizer;
+
+pub struct Analyzer<'a, A> {
+    config: AnalyzerConfig<'a, A>,
+}
+
+impl<'a, A> Analyzer<'a, A>
+where
+    A: AsRef<[u8]>
+{
+    /// create a new tokenizer detecting script
+    /// and chose the specialized internal tokenizer
+    pub fn new(config: AnalyzerConfig<'a, A>) -> Self { unimplemented!() }
+
+    /// Builds an `AnalyzedText` instance with the correct analyzer pipeline, and pre-processes the
+    /// text.
+    ///
+    /// If an analysis pipeline exists for the inferred `(Script, Language)`, the analyzer will look
+    /// for a user specified default `(Script::Other, Language::Other)`. If the user default is not
+    /// specified, it will fallback to `(IdentityPreProcessor, UnicodeSegmenter, IdentityNormalizer)`.
+    ///
+    /// ```rust
+    /// use meilisearch_tokenizer::{Analyzer, AnalyzerConfig};
+    /// use fst::Set;
+    /// // defaults to unicode segmenter with identity preprocessor and normalizer.
+    /// let stop_words = Set::from_iter([""].iter()).unwrap();
+    /// let analyzer = Analyzer::new(AnalyzerConfig::default_with_stopwords(&stop_words));
+    /// let analyzed = analyzer.analyze("The quick (\"brown\") fox can't jump 32.3 feet, right? Brr, it's 29.3°F!");
+    /// let mut tokens = analyzed.tokens();
+    /// assert!("the" == tokens.next().unwrap().text());
+    /// ```
+    pub fn analyze<'t>(&'t self, text: &'t str) -> AnalyzedText<'t, A> { unimplemented!() }
+}
+
+/// result of analyzer.analyze() function, made to choose how tokens will be returned.
+pub struct AnalyzedText<'a, A> {}
+
+impl<'a, A> AnalyzedText<'a, A>
+where
+    A: AsRef<[u8]>
+{
+    /// Returns a `TokenStream` for the Analyzed text.
+    pub fn tokens(&'a self) -> TokenStream<'a> { unimplemented!() }
+
+    /// Attaches each token to its corresponding portion of the original text.
+    pub fn reconstruct(&'a self) -> impl Iterator<Item = (&'a str, Token<'a>)>  { unimplemented!() }
+}
+```
+
+#### Token
+
+```rust
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum SeparatorKind {
+    Hard,
+    Soft,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum TokenKind {
+    Word,
+    /// the token is a stop word,
+    /// meaning that it can be ignored to optimize size and performance or be indexed as a Word
+    StopWord,
+    /// the token is a separator,
+    /// meaning that it shouldn't be indexed but used to determine word proximity
+    Separator(SeparatorKind),
+    Any,
```

Comment:
Is it a well chosen word? Wouldn't `Other` or `Unknown` be better names?

Reply:
it's true that `Other` would be better because `Any` suggests that the Token could be a `Word` or a `Separator`. `Other` suggests that it's none of the previous cases.
But, in our implementation of the tokenizer, we define a `Word` as a Token which is not a `Separator` or a `StopWord`. The `Any` case is here because the creation of the token and his classification is not made at the same time, and between these two steps we set the `token_kind` to `Any`. The final user of the tokenizer should never have a token set to `Any` unless the classification is skipped.

Reply:
```suggestion
    Unknown,
```
@MarinPostma, what do you think about using `Unknown` instead of `Any`? Which is clearer.